### ISSUE TITLE
bfac hang

### ISSUE THREAD

**Issue by leesoh** [mazen160/bfac#8]
First off thanks for putting the work into this tool. I'm unfortunately running into an issue where it stops working completely. I'm running on MacOS and am proxying through Burp to deal with a somewhat complicated authentication mechanism, so unproxied isn't really an option.

I've relaunched it several times from both a container and in a venv, but the behaviour remains the same. I haven't been able to find anything special about the requests it gets stuck on either, aside from the fact that it is always the 11th URL.

I'll do a bit more poking around to see whether I can find out what's going on but wanted to open an issue in case you'd run into this before.

**Comment by leesoh** [comment]
I threw a trace on it tonight and it seemed to get stuck looping on this:

```
--- modulename: threading, funcname: active_count
threading.py(1227):     with _active_limbo_lock:
threading.py(1228):         return len(_active) + len(_limbo)
bfac(1362):             time.sleep(0.5)
bfac(1327):             i.queue.clear()
bfac(1326):             time.sleep(1)
bfac(1327):             i.queue.clear()
bfac(1326):             time.sleep(1)
bfac(1335):     while current_scan_queue.empty() is False:
 --- modulename: queue, funcname: empty
queue.py(107):         with self.mutex:
queue.py(108):             return not self._qsize()
 --- modulename: queue, funcname: _qsize
queue.py(209):         return len(self.queue)
bfac(1337):         if threading.active_count() < (int(max_threads) + 2):
```

**Comment by mazen160** [comment]
Hi @leesoh 

Thank you very much for your comprehensive check regarding the issue. I remember I once encountered the same issue, but I wasn't able to find the reason behind it. I knew that the issue affected some handling part of threading. I added a print statement to constantly print active threads during running the app. I noticed that there is a single thread that remains active without shutting down. The app reaches a deadlock at 11-13th URL in my case.

Further debugging showed me that the issue is at the `counter_reset` sub-function that acts as a request rate throttling functionality to evade bots detection.

I fixed the bug by properly managing the function and by adding it to threads state handler.
